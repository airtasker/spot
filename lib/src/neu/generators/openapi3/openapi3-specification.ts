// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#openapi-object
export interface OpenApiV3 {
  openapi: "3.0.2";
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObject;
  components?: ComponentsObject;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#info-object
export interface InfoObject {
  title: string;
  description?: string;
  termsOfService?: string;
  contact?: ContactObject;
  license?: LicenseObject;
  version: string;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#contact-object
export interface ContactObject {
  name?: string;
  url?: string;
  email?: string;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#license-object
export interface LicenseObject {
  name: string;
  url?: string;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#server-object
export interface ServerObject {
  url: string;
  description?: string;
  variables?: { [serverVariable: string]: ServerVariableObject };
}

// https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md#serverVariableObject
export interface ServerVariableObject {
  enum?: string[];
  default: string;
  description?: string;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#pathsObject
export interface PathsObject {
  [path: string]: PathItemObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#pathItemObject
export interface PathItemObject {
  $ref?: string;
  summary?: string;
  description?: string;
  get?: OperationObject;
  put?: OperationObject;
  post?: OperationObject;
  delete?: OperationObject;
  options?: OperationObject;
  head?: OperationObject;
  patch?: OperationObject;
  trace?: OperationObject;
  servers?: ServerObject[];
  parameters?: Array<ParameterObject | ReferenceObject>;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#operationObject
export interface OperationObject {
  tags?: string[];
  summary?: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
  operationId?: string;
  parameters?: Array<ParameterObject | ReferenceObject>;
  requestBody?: RequestBodyObject | ReferenceObject;
  responses: ResponsesObject;
  callbacks?: { [callback: string]: CallbackObject | ReferenceObject };
  deprecated?: boolean;
  security?: SecurityRequirementObject[];
  servers?: ServerObject[];
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#referenceObject
export interface ReferenceObject {
  $ref: string;
  /**
   * WARNING
   *
   * `nullable: true` will occur when exactly one type reference is combined with null.
   *
   * Example:
   *
   * `MyType | null`
   *
   * https://swagger.io/docs/specification/using-ref/#considerations
   *
   * Schema references cannot contain sibling elements. `nullable` therefore should
   * not be combined with schema reference objects. This rule was misunderstood during
   * development on the OpenAPI 3 generator at Airtasker. This will be removed in a
   * future version of Spot when Airtasker's tooling supports an alternative valid
   * representation for the above scenario.
   *
   * TODO: Find a way to remove this
   * A possible seemingly accepted workaround to this is to wrap the schema reference
   * into an allOf.
   *
   * Example:
   *
   * ```
   * nullable: true
   * allOf:
   *   - $ref: #/components/schemas/MyType
   * ```
   */
  nullable?: boolean;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#requestBodyObject
export interface RequestBodyObject {
  description?: string;
  content: { [mediaType: string]: MediaTypeObject };
  required?: boolean;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#mediaTypeObject
export type MediaTypeObject = {
  schema?: SchemaObject | ReferenceObject;
  encoding?: { [encoding: string]: EncodingObject };
} & MutuallyExclusiveExample;

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schemaObject
export type SchemaObject =
  | NumberSchemaObject
  | IntegerSchemaObject
  | StringSchemaObject
  | BooleanSchemaObject
  | ArraySchemaObject
  | ObjectSchemaObject
  | AnySchemaObject
  | AllOfSchemaObject
  | OneOfSchemaObject
  | AnyOfSchemaObject
  | NotSchemaObject;

interface SchemaObjectBase {
  nullable?: boolean;
  not?: SchemaObject | ReferenceObject;
  title?: string;
  description?: string;
  example?: any;
  externalDocs?: ExternalDocumentationObject;
  deprecated?: boolean;
}

export interface NumberSchemaObject
  extends SchemaObjectBase,
    NumberSchemaObjectBase {
  type: "number";
  format?: "float" | "double";
}

export interface IntegerSchemaObject
  extends SchemaObjectBase,
    NumberSchemaObjectBase {
  type: "integer";
  format?: "int32" | "int64";
}

interface NumberSchemaObjectBase {
  maximum?: number;
  exclusiveMaximum?: boolean;
  minimum?: number;
  exclusiveMinimum?: boolean;
  multipleOf?: number;
  enum?: Array<number | null>;
  default?: number;
}

export interface StringSchemaObject extends SchemaObjectBase {
  type: "string";
  maxLength?: number;
  minLength?: number;
  /**
   * OpenAPI allows custom formats. We constrain the format here to those
   * that OpenAPI has defined and custom formats that Spot may produce.
   */
  format?: "date" | "date-time" | "password" | "byte" | "binary";
  pattern?: string;
  enum?: Array<string | null>;
  default?: string;
}

export interface BooleanSchemaObject extends SchemaObjectBase {
  type: "boolean";
  enum?: Array<boolean | null>;
  default?: boolean;
}

export interface ArraySchemaObject extends SchemaObjectBase {
  type: "array";
  items: SchemaObject | ReferenceObject;
  minItems?: number;
  maxItems?: number;
  uniqueItems?: boolean;
  default?: any[];
}

export interface ObjectSchemaObject extends SchemaObjectBase {
  type: "object";
  properties?: ObjectPropertiesSchemaObject;
  required?: string[];
  additionalProperties?: SchemaObject | ReferenceObject | boolean;
  maxProperties?: number;
  minProperties?: number;
  default?: any;
}

export interface ObjectPropertiesSchemaObject {
  [name: string]:
    | (SchemaObject & ObjectPropertySchemaObjectBase)
    | ReferenceObject;
}

interface ObjectPropertySchemaObjectBase {
  xml?: XmlObject;
  readOnly?: boolean;
  writeOnly?: boolean;
}

export interface AnySchemaObject extends SchemaObjectBase {
  AnyValue: {};
}

export interface AllOfSchemaObject extends SchemaObjectBase {
  allOf: Array<SchemaObject | ReferenceObject>;
  discriminator?: DiscriminatorObject;
}

export interface OneOfSchemaObject extends SchemaObjectBase {
  oneOf: Array<SchemaObject | ReferenceObject>;
  discriminator?: DiscriminatorObject;
}

export interface AnyOfSchemaObject extends SchemaObjectBase {
  anyOf: Array<SchemaObject | ReferenceObject>;
  discriminator?: DiscriminatorObject;
}

export interface NotSchemaObject extends SchemaObjectBase {
  not: SchemaObject | ReferenceObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#discriminatorObject
export interface DiscriminatorObject {
  propertyName: string;
  mapping?: { [key: string]: string };
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#securitySchemeObject
export type SecuritySchemeObject =
  | ApiKeySecuritySchemeObject
  | HttpSecuritySchemeObject
  | OAuth2SecuritySchemeObject
  | OpenIdConnectSecuritySchemeObject;

export interface ApiKeySecuritySchemeObject extends SecuritySchemeObjectBase {
  type: "apiKey";
  name: string;
  in: "query" | "header" | "cookie";
}

export interface HttpSecuritySchemeObject extends SecuritySchemeObjectBase {
  type: "http";
  scheme: string;
  bearerFormat?: string;
}

export interface OAuth2SecuritySchemeObject extends SecuritySchemeObjectBase {
  type: "oauth2";
  flows: OAuthFlowsObject;
}

export interface OpenIdConnectSecuritySchemeObject
  extends SecuritySchemeObjectBase {
  type: "openIdConnect";
  openIdConnectUrl: string;
}

interface SecuritySchemeObjectBase {
  type: SecuritySchemeType;
  description?: string;
}

type SecuritySchemeType = "apiKey" | "http" | "oauth2" | "openIdConnect";

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#oauthFlowsObject
export interface OAuthFlowsObject {
  implicit?: ImplicitOAuthFlowObject;
  password?: PasswordOAuthFlowObject;
  clientCredentials?: ClientCredentialsOAuthFlowObject;
  authorizationCode?: AuthorizationCodeOAuthFlowObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#oauthFlowObject
export interface ImplicitOAuthFlowObject extends OAuthFlowObjectBase {
  authorizationUrl: string;
}

export interface PasswordOAuthFlowObject extends OAuthFlowObjectBase {
  tokenUrl: string;
}

export interface ClientCredentialsOAuthFlowObject extends OAuthFlowObjectBase {
  tokenUrl: string;
}

export interface AuthorizationCodeOAuthFlowObject extends OAuthFlowObjectBase {
  authorizationUrl: string;
  tokenUrl: string;
}

interface OAuthFlowObjectBase {
  refreshUrl?: string;
  scopes: { [scope: string]: string };
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#securityRequirementObject
export interface SecurityRequirementObject {
  [name: string]: string[];
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#componentsObject
export interface ComponentsObject {
  schemas?: { [schema: string]: SchemaObject | ReferenceObject };
  responses?: { [response: string]: ResponseObject | ReferenceObject };
  parameters?: { [parameter: string]: ParameterObject | ReferenceObject };
  examples?: { [example: string]: ExampleObject | ReferenceObject };
  requestBodies?: { [request: string]: RequestBodyObject | ReferenceObject };
  headers?: { [header: string]: HeaderObject | ReferenceObject };
  securitySchemes?: {
    [securityScheme: string]: SecuritySchemeObject | ReferenceObject;
  };
  links?: { [link: string]: LinkObject | ReferenceObject };
  callbacks?: { [callback: string]: CallbackObject | ReferenceObject };
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#xmlObject
export interface XmlObject {
  name?: string;
  namespace?: string;
  prefix?: string;
  attribute?: boolean;
  wrapped?: boolean;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#exampleObject
export type ExampleObject = {
  summary?: string;
  description?: string;
  value?: any;
  externalValue?: string;
} & MutuallyExclusiveExampleObjectValue;

type MutuallyExclusiveExampleObjectValue =
  | {
      value: any;
      externalValue?: never;
    }
  | {
      value?: never;
      externalValue: string;
    };

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#encoding-object
export interface EncodingObject {
  contentType?: string;
  headers?: { [name: string]: HeaderObject | ReferenceObject };
  style?: string;
  explode?: boolean;
  allowReserved?: boolean;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#responsesObject
export interface ResponsesObject {
  [statusCodeOrDefault: string]: ResponseObject | ReferenceObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#responseObject
export interface ResponseObject {
  description: string;
  headers?: { [name: string]: HeaderObject | ReferenceObject };
  content?: { [mediaType: string]: MediaTypeObject };
  links?: { [link: string]: LinkObject | ReferenceObject };
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#parameterObject
export type ParameterObject =
  | QueryParameterObject
  | HeaderParameterObject
  | PathParameterObject
  | CookieParameterObject;

export type QueryParameterObject = ParameterObjectBase & {
  in: "query";
  allowEmptyValue?: boolean;
  style?: "form" | "spaceDelimited" | "pipeDelimited" | "deepObject";
  allowReserved?: boolean;
};

export type HeaderParameterObject = ParameterObjectBase & {
  in: "header";
  style?: "simple";
};

export type PathParameterObject = ParameterObjectBase & {
  in: "path";
  required: true;
  style?: "simple" | "label" | "matrix";
};

export type CookieParameterObject = ParameterObjectBase & {
  in: "cookie";
  style?: "form";
};

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#headerObject
export type HeaderObject = Omit<HeaderParameterObject, "name" | "in">;

type ParameterObjectBase = {
  name: string;
  in: "query" | "header" | "path" | "cookie";
  description?: string;
  required?: boolean;
  deprecated?: boolean;

  style?: ParameterStyle;
  explode?: boolean;
  schema?: SchemaObject | ReferenceObject;
} & MutuallyExclusiveExample;

type ParameterStyle =
  | "matrix"
  | "label"
  | "form"
  | "simple"
  | "spaceDelimited"
  | "pipeDelimited"
  | "deepObject";

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#callbackObject
export interface CallbackObject {
  [name: string]: PathItemObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#linkObject
type LinkObject = {
  parameters?: { [name: string]: any };
  requestBody?: any;
  description?: string;
  server?: ServerObject;
} & MutuallyExclusiveLinkObjectOperation;

type MutuallyExclusiveLinkObjectOperation =
  | {
      operationRef: string;
      operationId?: never;
    }
  | {
      operationRef?: never;
      operationId: string;
    };

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#tagObject
export interface TagObject {
  name: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
}

// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#externalDocumentationObject
export interface ExternalDocumentationObject {
  description?: string;
  url: string;
}

// Common

type MutuallyExclusiveExample =
  | {
      example?: any;
      examples?: never;
    }
  | {
      example?: never;
      examples?: { [example: string]: ExampleObject | ReferenceObject };
    };
