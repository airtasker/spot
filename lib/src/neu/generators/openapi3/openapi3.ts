import assertNever from "assert-never";
import {
  Body,
  Config,
  Contract,
  DefaultResponse,
  Endpoint,
  Header,
  HttpMethod,
  isSpecificResponse,
  Request,
  Response
} from "../../definitions";
import {
  isArrayType,
  isNotNullType,
  isObjectType,
  possibleRootTypes,
  Type,
  TypeTable
} from "../../types";
import { KeyOfType } from "../../util";
import {
  ComponentsObject,
  HeaderObject,
  HeaderParameterObject,
  OpenApiV3,
  OperationObject,
  ParameterObject,
  PathItemObject,
  PathParameterObject,
  PathsObject,
  QueryParameterObject,
  ReferenceObject,
  RequestBodyObject,
  ResponseObject,
  ResponsesObject,
  SchemaObject
} from "./openapi3-schema";
import { typeToSchemaOrReferenceObject } from "./openapi3-type-util";

const SECURITY_HEADER_SCHEME_NAME = "SecurityHeader";

function generateOpenAPI3(contract: Contract): OpenApiV3 {
  const typeTable = TypeTable.fromArray(contract.types);
  const openapi: OpenApiV3 = {
    openapi: "3.0.2",
    info: {
      title: contract.name,
      description: contract.description,
      version: "0.0.0"
    },
    paths: endpointsToPathsObject(
      contract.endpoints,
      typeTable,
      contract.config
    ),
    components: {
      schemas: contractTypesToComponentsObjectSchemas(
        contract.types,
        typeTable
      ),
      securitySchemes: contract.security && {
        [SECURITY_HEADER_SCHEME_NAME]: {
          type: "apiKey",
          in: "header",
          name: contract.security.name,
          description: contract.security.description
        }
      }
    },
    security: contract.security && [
      {
        [SECURITY_HEADER_SCHEME_NAME]: []
      }
    ]
  };

  return openapi;
}

function endpointsToPathsObject(
  endpoints: Endpoint[],
  typeTable: TypeTable,
  config: Config
): PathsObject {
  return endpoints.reduce<PathsObject>((acc, endpoint) => {
    acc[endpoint.path] = acc[endpoint.path] || {};
    const pathItemMethod = httpMethodToPathItemMethod(endpoint.method);
    acc[endpoint.path][pathItemMethod] = endpointToOperationObject(
      endpoint,
      typeTable,
      config
    );
    return acc;
  }, {});
}

function endpointToOperationObject(
  endpoint: Endpoint,
  typeTable: TypeTable,
  config: Config
): OperationObject {
  const endpointRequest = endpoint.request;
  const endpointRequestBody = endpointRequest && endpointRequest.body;

  return {
    tags: endpoint.tags,
    description: endpoint.description,
    operationId: endpoint.name,
    parameters:
      endpointRequest &&
      endpointRequestToParameterObjects(endpointRequest, typeTable, config),
    requestBody:
      endpointRequestBody &&
      endpointRequestBodyToRequestBodyObject(endpointRequestBody, typeTable),
    responses: endpointResponsesToResponsesObject(
      {
        specific: endpoint.responses,
        default: endpoint.defaultResponse
      },
      typeTable
    )
  };
}

function endpointRequestToParameterObjects(
  request: Request,
  typeTable: TypeTable,
  config: Config
): ParameterObject[] {
  const pathParameters: PathParameterObject[] = request.pathParams.map(p => {
    return {
      name: p.name,
      in: "path",
      description: p.description,
      required: true,
      schema: typeToSchemaOrReferenceObject(p.type, typeTable)
    };
  });

  const queryParameters: QueryParameterObject[] = request.queryParams.map(p => {
    return {
      name: p.name,
      in: "query",
      description: p.description,
      ...typeToQueryParameterSerializationStrategy(p.type, typeTable, config),
      required: !p.optional,
      schema: typeToSchemaOrReferenceObject(p.type, typeTable)
    };
  });

  const headerParameters: HeaderParameterObject[] = request.headers.map(p => {
    return {
      name: p.name,
      in: "header",
      description: p.description,
      required: !p.optional,
      schema: typeToSchemaOrReferenceObject(p.type, typeTable)
    };
  });

  const parameters: ParameterObject[] = [];

  return parameters.concat(pathParameters, queryParameters, headerParameters);
}

function typeToQueryParameterSerializationStrategy(
  type: Type,
  typeTable: TypeTable,
  config: Config
): Pick<QueryParameterObject, "style" | "explode"> {
  const possibleTypes = possibleRootTypes(type, typeTable).filter(
    isNotNullType
  );

  if (possibleTypes.length === 0) {
    throw new Error("Unexpected error: query param resolved to no types");
  }

  const possiblyObjectType = possibleTypes.some(isObjectType);
  const possiblyArrayType = possibleTypes.some(isArrayType);

  if (possiblyObjectType && !possiblyArrayType) {
    return { style: "deepObject", explode: true };
  }

  if (possiblyArrayType && !possiblyObjectType) {
    switch (config.paramSerializationStrategy.query.array) {
      case "ampersand": {
        return { style: "form", explode: true };
      }
      case "comma": {
        return { style: "form", explode: false };
      }
      default:
        assertNever(config.paramSerializationStrategy.query.array);
    }
  }

  return {};
}

function endpointRequestBodyToRequestBodyObject(
  requestBody: Body,
  typeTable: TypeTable
): RequestBodyObject {
  const content = {
    "application/json": {
      schema: typeToSchemaOrReferenceObject(requestBody.type, typeTable)
    }
  };

  // TODO: currently Spot does not support optional request body
  return { content, required: true };
}

function endpointResponsesToResponsesObject(
  responses: {
    specific: Response[];
    default?: DefaultResponse;
  },
  typeTable: TypeTable
): ResponsesObject {
  const responsesObject = responses.specific.reduce<ResponsesObject>(
    (acc, response) => {
      acc[response.status.toString(10)] = endpointResponseToResponseObject(
        response,
        typeTable
      );
      return acc;
    },
    {}
  );

  if (responses.default) {
    responsesObject.default = endpointResponseToResponseObject(
      responses.default,
      typeTable
    );
  }

  return responsesObject;
}

function endpointResponseToResponseObject(
  response: Response | DefaultResponse,
  typeTable: TypeTable
): ResponseObject {
  const description =
    response.description ||
    (isSpecificResponse(response)
      ? `${response.status} response`
      : "default response");

  const headers =
    response.headers.length > 0
      ? response.headers.reduce<{ [name: string]: HeaderObject }>(
          (acc, header) => {
            acc[header.name] = headerToHeaderObject(header, typeTable);
            return acc;
          },
          {}
        )
      : undefined;

  const content = response.body && {
    "application/json": {
      schema: typeToSchemaOrReferenceObject(response.body.type, typeTable)
    }
  };

  return { description, headers, content };
}

function headerToHeaderObject(
  header: Header,
  typeTable: TypeTable
): HeaderObject {
  return {
    description: header.description,
    required: !header.optional,
    schema: typeToSchemaOrReferenceObject(header.type, typeTable)
  };
}

function contractTypesToComponentsObjectSchemas(
  types: Contract["types"],
  typeTable: TypeTable
): ComponentsObject["schemas"] {
  return types.reduce<{
    [schema: string]: SchemaObject | ReferenceObject;
  }>((acc, t) => {
    acc[t.name] = typeToSchemaOrReferenceObject(t.type, typeTable);
    return acc;
  }, {});
}

function httpMethodToPathItemMethod(
  method: HttpMethod
): KeyOfType<PathItemObject, OperationObject> {
  switch (method) {
    case "GET":
      return "get";
    case "PUT":
      return "put";
    case "POST":
      return "post";
    case "DELETE":
      return "delete";
    case "PATCH":
      return "patch";
    default:
      assertNever(method);
  }
}
