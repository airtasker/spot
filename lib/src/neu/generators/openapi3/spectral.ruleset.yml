# https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/reference/openapi-rules.md
extends: [[spectral:oas, off]]
rules:
  operation-2xx-response: true
  operation-operationId-unique: true
  operation-parameters: true
  path-params: true
  example-value-or-externalValue: true
  no-eval-in-markdown: true
  no-script-tags-in-markdown: true
  openapi-tags-alphabetical: true
  operation-operationId-valid-in-url: true
  path-declarations-must-exist: true
  path-keys-no-trailing-slash: true
  path-not-include-query: true
  oas3-operation-security-defined: true
  oas3-server-not-example.com: true
  oas3-server-trailing-slash: true
  oas3-unused-components-schema: true
  oas3-valid-example: true
  oas3-schema: true
