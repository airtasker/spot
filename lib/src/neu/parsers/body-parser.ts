import { ParameterDeclaration } from "ts-morph";
import { Body } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseType } from "./type-parser";

export function parseBody(
  parameter: ParameterDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): Body {
  const decorator = parameter.getDecoratorOrThrow("body");
  if (parameter.hasQuestionToken()) {
    throw new Error("@body parameter cannot be optional");
  }
  const type = parseType(parameter.getTypeNodeOrThrow(), typeTable, lociTable);
  // TODO: add loci information
  return {
    // TODO: how to extract description from parameter declaration?
    description: undefined,
    type
  };
}
