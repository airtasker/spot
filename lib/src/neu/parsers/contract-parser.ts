import { ClassDeclaration, SourceFile } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Config, Contract } from "../definitions";
import { ParserError } from "../errors";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { ok, Result } from "../util";
import { defaultConfig, parseConfig } from "./config-parser";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

/**
 * Parse a root source file to return a contract.
 */
export function parseContract(
  file: SourceFile
): Result<{ contract: Contract; lociTable: LociTable }, Error> {
  const typeTable = new TypeTable();
  const lociTable = new LociTable();

  const klass = getClassWithDecoratorOrThrow(file, "api"); // TODO: throw a custom error
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  // Handle config
  const configResult = resolveConfig(klass);
  if (configResult.isErr()) return configResult;

  // Handle security
  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");
  const security =
    securityHeaderProp &&
    parseSecurityHeader(securityHeaderProp, typeTable, lociTable);
  if (security && security.isErr()) return security;

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpointClasses = projectFiles.reduce<ClassDeclaration[]>(
    (acc, currentFile) =>
      acc.concat(
        currentFile
          .getClasses()
          .filter(k => k.getDecorator("endpoint") !== undefined)
      ),
    []
  );
  const endpoints = [];
  for (const k of endpointClasses) {
    const endpointResult = parseEndpoint(k, typeTable, lociTable);
    if (endpointResult.isErr()) return endpointResult;
    endpoints.push(endpointResult.unwrap());
  }

  const contract = {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    config: configResult.unwrap(),
    security: security && security.unwrap(),
    endpoints: endpoints.sort((a, b) => (b.name > a.name ? -1 : 1))
  };

  return ok({ contract, lociTable });
}

function resolveConfig(klass: ClassDeclaration): Result<Config, ParserError> {
  const hasConfigDecorator = klass.getDecorator("config") !== undefined;
  if (hasConfigDecorator) {
    return parseConfig(klass);
  } else {
    return ok(defaultConfig());
  }
}
