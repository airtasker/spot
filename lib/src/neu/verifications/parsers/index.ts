// @ts-ignore
import * as antlr4 from 'antlr4';
// @ts-ignore
import { QueryParamsLexer } from './QueryParamsLexer';
// @ts-ignore
import { QueryParamsParser } from './QueryParamsParser';

// let MyGrammarParser = require('./QueryParamsLexer').MyGrammarParser;
// let MyGrammarListener = require('./QueryParamsLexer').MyGrammarListener;

export function parseQueryParams(input:string) {
    let chars = new antlr4.InputStream(input);
    let lexer = new QueryParamsLexer(chars);
    let tokens  = new antlr4.CommonTokenStream(lexer);
    // console.log(tokens);
    let parser = new QueryParamsParser(tokens);
    parser.buildParseTrees = true;
    let tree = parser.querystring();
    console.log(tree.toStringTree(parser.ruleNames)); 
}
