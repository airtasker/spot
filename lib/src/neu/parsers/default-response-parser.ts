import { MethodDeclaration } from "ts-morph";
import { DefaultResponse } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseBody } from "./body-parser";
import { parseHeaders } from "./headers-parser";
import { getJsDoc, getParamWithDecorator } from "./parser-helpers";

export function parseDefaultResponse(
  method: MethodDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): DefaultResponse {
  method.getDecoratorOrThrow("defaultResponse");
  const headersParam = getParamWithDecorator(method, "headers");
  const headers = headersParam
    ? parseHeaders(headersParam, typeTable, lociTable)
    : [];
  const bodyParam = getParamWithDecorator(method, "body");
  const body = bodyParam && parseBody(bodyParam, typeTable, lociTable);
  const descriptionDoc = getJsDoc(method);
  return {
    headers,
    description: descriptionDoc && descriptionDoc.getComment(),
    body
  };
}
