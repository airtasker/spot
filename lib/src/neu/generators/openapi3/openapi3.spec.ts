import { createProjectFromExistingSourceFile } from "../../../spec-helpers/helper";
import { parseContract } from "../../parsers/contract-parser";
import { generateOpenAPI3 } from "./openapi3";

describe("OpenAPI 3 generator", () => {
  test("produces valid code", () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/contract.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateOpenAPI3(contract);

    expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
  });
});
