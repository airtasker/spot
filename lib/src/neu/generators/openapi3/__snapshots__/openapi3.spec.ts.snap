// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI 3 generator produces valid code 1`] = `
"{
  \\"openapi\\": \\"3.0.2\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"description\\": \\"contract description\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"paths\\": {
    \\"/path/:param/nest\\": {
      \\"get\\": {
        \\"tags\\": [],
        \\"operationId\\": \\"GetEndpoint\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"param\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"string\\"
            }
          },
          {
            \\"name\\": \\"param\\",
            \\"in\\": \\"query\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"string\\"
            }
          },
          {
            \\"name\\": \\"header\\",
            \\"in\\": \\"header\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"string\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"headers\\": {
              \\"responseHeader\\": {
                \\"required\\": true,
                \\"schema\\": {
                  \\"type\\": \\"string\\"
                }
              }
            },
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/SuccessBody\\"
                }
              }
            }
          },
          \\"default\\": {
            \\"description\\": \\"default response\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/DefaultBody\\"
                }
              }
            }
          }
        }
      }
    }
  },
  \\"components\\": {
    \\"schemas\\": {
      \\"DefaultBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"message\\": {
            \\"type\\": \\"string\\"
          }
        }
      },
      \\"SuccessBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"message\\": {
            \\"type\\": \\"string\\"
          }
        }
      }
    },
    \\"securitySchemes\\": {
      \\"SecurityHeader\\": {
        \\"type\\": \\"apiKey\\",
        \\"in\\": \\"header\\",
        \\"name\\": \\"security-header\\"
      }
    }
  },
  \\"security\\": [
    {
      \\"SecurityHeader\\": []
    }
  ]
}"
`;
